File: R/scoring.R
clamp_score <- function(x, lower = 0, upper = 10) {
  pmax(lower, pmin(upper, x))
}

scenario_multipliers <- function(scenarios, scenario_name, categories) {
  if (!(scenario_name %in% scenarios$scenario)) {
    available <- paste(sort(unique(scenarios$scenario)), collapse = ", ")
    stop(sprintf("Unknown scenario '%s'. Available scenarios: %s", scenario_name, available), call. = FALSE)
  }

  scenario_rows <- scenarios[scenarios$scenario == scenario_name, , drop = FALSE]
  multipliers <- setNames(rep(1, length(categories)), categories)
  matched <- match(categories, scenario_rows$category)
  has_match <- !is.na(matched)
  multipliers[has_match] <- scenario_rows$multiplier[matched[has_match]]
  multipliers
}

prepare_compare_frame <- function(criteria, weights, options, scenarios, scenario_name) {
  active_criteria <- criteria[criteria$active, , drop = FALSE]
  options <- resolve_option_scores(options, criteria)
  schools <- default_schools()[default_schools() %in% unique(options$school)]

  if (length(schools) != 2) {
    stop("The comparison workflow currently expects exactly two schools.", call. = FALSE)
  }

  synced_weights <- sync_weights(weights, active_criteria)
  multipliers <- scenario_multipliers(scenarios, scenario_name, unique(active_criteria$category))

  base_frame <- active_criteria[, c("criterion_id", "category", "label", "description", "default_weight")]
  base_frame <- merge(base_frame, synced_weights, by = "criterion_id", all.x = TRUE, sort = FALSE)
  base_frame$effective_weight <- base_frame$user_weight * multipliers[base_frame$category]

  school_a <- options[options$school == schools[[1]], c("criterion_id", "score", "confidence", "dealbreaker", "note", "known_fact")]
  school_b <- options[options$school == schools[[2]], c("criterion_id", "score", "confidence", "dealbreaker", "note", "known_fact")]

  names(school_a) <- c("criterion_id", "score_a", "confidence_a", "dealbreaker_a", "note_a", "known_fact_a")
  names(school_b) <- c("criterion_id", "score_b", "confidence_b", "dealbreaker_b", "note_b", "known_fact_b")

  frame <- merge(base_frame, school_a, by = "criterion_id", all.x = TRUE, sort = FALSE)
  frame <- merge(frame, school_b, by = "criterion_id", all.x = TRUE, sort = FALSE)

  frame$paired <- !is.na(frame$score_a) & !is.na(frame$score_b)
  frame$confidence_a[is.na(frame$confidence_a)] <- 0.5
  frame$confidence_b[is.na(frame$confidence_b)] <- 0.5
  frame$adjusted_score_a <- clamp_score(frame$score_a - 1.5 * (1 - frame$confidence_a))
  frame$adjusted_score_b <- clamp_score(frame$score_b - 1.5 * (1 - frame$confidence_b))
  frame$score_gap <- frame$adjusted_score_a - frame$adjusted_score_b
  frame$contribution_gap <- frame$effective_weight * frame$score_gap
  frame
}

weighted_score <- function(scores, weights) {
  valid <- !is.na(scores) & !is.na(weights) & weights > 0
  if (!any(valid)) {
    return(NA_real_)
  }
  weighted.mean(scores[valid], weights[valid])
}

category_summary <- function(frame, schools) {
  categories <- unique(frame$category)
  out <- list()

  for (cat_name in categories) {
    subset_rows <- frame[frame$category == cat_name, , drop = FALSE]
    paired_rows <- subset_rows[subset_rows$paired, , drop = FALSE]
    total_criteria <- nrow(subset_rows)
    criteria_scored <- nrow(paired_rows)

    out[[length(out) + 1]] <- data.frame(
      category = cat_name,
      school = schools[[1]],
      score = weighted_score(paired_rows$adjusted_score_a, paired_rows$effective_weight),
      criteria_scored = criteria_scored,
      total_criteria = total_criteria,
      stringsAsFactors = FALSE
    )
    out[[length(out) + 1]] <- data.frame(
      category = cat_name,
      school = schools[[2]],
      score = weighted_score(paired_rows$adjusted_score_b, paired_rows$effective_weight),
      criteria_scored = criteria_scored,
      total_criteria = total_criteria,
      stringsAsFactors = FALSE
    )
  }

  do.call(rbind, out)
}

dealbreaker_warnings <- function(options, criteria, threshold = 6) {
  options <- resolve_option_scores(options, criteria)
  merged <- merge(
    options,
    criteria[, c("criterion_id", "label")],
    by = "criterion_id",
    all.x = TRUE,
    sort = FALSE
  )

  merged$dealbreaker <- normalize_logical(merged$dealbreaker)
  warnings <- merged[merged$dealbreaker & !is.na(merged$score) & merged$score < threshold, c("school", "criterion_id", "label", "score", "note")]
  warnings[order(warnings$school, warnings$score), , drop = FALSE]
}

compare_schools <- function(criteria, weights, options, scenarios, scenario_name = "balanced") {
  options <- resolve_option_scores(options, criteria)
  frame <- prepare_compare_frame(criteria, weights, options, scenarios, scenario_name)
  schools <- default_schools()[default_schools() %in% unique(options$school)]
  paired_rows <- frame[frame$paired, , drop = FALSE]

  overall <- data.frame(
    school = schools,
    overall_score = c(
      weighted_score(paired_rows$adjusted_score_a, paired_rows$effective_weight),
      weighted_score(paired_rows$adjusted_score_b, paired_rows$effective_weight)
    ),
    stringsAsFactors = FALSE
  )

  drivers <- paired_rows[order(-abs(paired_rows$contribution_gap)), c("criterion_id", "label", "score_gap")]
  if (nrow(drivers) > 0) {
    drivers$leading_school <- ifelse(drivers$score_gap >= 0, schools[[1]], schools[[2]])
    drivers$score_gap <- abs(drivers$score_gap)
    drivers <- head(drivers, 10)
  }

  missing <- frame[!frame$paired, c("criterion_id", "label", "category", "effective_weight")]

  list(
    scenario = scenario_name,
    schools = schools,
    frame = frame,
    coverage = if (nrow(frame) == 0) 0 else nrow(paired_rows) / nrow(frame),
    overall = overall,
    category_summary = category_summary(frame, schools),
    drivers = drivers,
    missing = missing,
    dealbreakers = dealbreaker_warnings(options, criteria)
  )
}

File: src/lib/scoring.js
import { CATEGORY_COPY, getPlainExplanation } from './content.js'
import { SCHOOL_ORDER } from './data.js'

function clampScore(value, minimum = 0, maximum = 10) {
  return Math.min(maximum, Math.max(minimum, value))
}

function weightedMean(items, valueKey, weightKey) {
  let numerator = 0
  let denominator = 0

  for (const item of items) {
    const value = item[valueKey]
    const weight = item[weightKey]
    if (value == null || weight == null || weight <= 0) {
      continue
    }
    numerator += value * weight
    denominator += weight
  }

  return denominator === 0 ? null : numerator / denominator
}

export function resolveOption(option) {
  if (option.scoreOwner === 'camille_required') {
    return {
      score: option.userScore ?? option.baselineScore,
      confidence: option.userConfidence ?? option.baselineConfidence,
      source: 'Survey',
      note: option.userNote || option.baselineNote,
    }
  }

  return {
    score: option.baselineScore,
    confidence: option.baselineConfidence,
    source: 'Research baseline',
    note: option.baselineNote,
  }
}

function buildDetailRows(criteria, weights, options) {
  const optionMap = new Map(options.map((option) => [option.key, option]))
  const weightMap = new Map(weights.map((weight) => [weight.criterionId, weight.userWeight]))

  return criteria.map((criterion) => {
    const ucsb = resolveOption(optionMap.get(`UCSB::${criterion.criterionId}`))
    const northwestern = resolveOption(optionMap.get(`Northwestern::${criterion.criterionId}`))
    return {
      criterionId: criterion.criterionId,
      category: criterion.category,
      categoryLabel: CATEGORY_COPY[criterion.category],
      label: criterion.label,
      description: criterion.description,
      plainExplanation: getPlainExplanation(criterion),
      weight: weightMap.get(criterion.criterionId) ?? criterion.defaultWeight,
      ucsbScore: ucsb.score,
      ucsbConfidence: ucsb.confidence,
      ucsbSource: ucsb.source,
      ucsbNote: ucsb.note,
      northwesternScore: northwestern.score,
      northwesternConfidence: northwestern.confidence,
      northwesternSource: northwestern.source,
      northwesternNote: northwestern.note,
      gap: ucsb.score - northwestern.score,
    }
  })
}

export function buildComparison(criteria, weights, options) {
  const detailRows = buildDetailRows(criteria, weights, options)
  const frame = detailRows.map((row) => {
    const adjustedUcsb = clampScore(row.ucsbScore - 1.5 * (1 - row.ucsbConfidence))
    const adjustedNorthwestern = clampScore(row.northwesternScore - 1.5 * (1 - row.northwesternConfidence))
    return {
      ...row,
      paired: row.ucsbScore != null && row.northwesternScore != null,
      adjustedUcsb,
      adjustedNorthwestern,
      scoreGap: adjustedUcsb - adjustedNorthwestern,
      contributionGap: row.weight * (adjustedUcsb - adjustedNorthwestern),
    }
  })

  const pairedRows = frame.filter((row) => row.paired)
  const overall = [
    { school: SCHOOL_ORDER[0], overallScore: weightedMean(pairedRows, 'adjustedUcsb', 'weight') },
    { school: SCHOOL_ORDER[1], overallScore: weightedMean(pairedRows, 'adjustedNorthwestern', 'weight') },
  ]

  const categorySummary = Object.values(
    frame.reduce((accumulator, row) => {
      accumulator[row.category] ??= {
        category: row.category,
        schoolScores: {
          UCSB: [],
          Northwestern: [],
        },
        totalCriteria: 0,
        criteriaScored: 0,
      }

      const bucket = accumulator[row.category]
      bucket.totalCriteria += 1
      if (row.paired) {
        bucket.criteriaScored += 1
        bucket.schoolScores.UCSB.push(row)
        bucket.schoolScores.Northwestern.push(row)
      }
      return accumulator
    }, {}),
  ).flatMap((bucket) => {
    return [
      {
        category: bucket.category,
        school: 'UCSB',
        score: weightedMean(bucket.schoolScores.UCSB, 'adjustedUcsb', 'weight'),
        criteriaScored: bucket.criteriaScored,
        totalCriteria: bucket.totalCriteria,
      },
      {
        category: bucket.category,
        school: 'Northwestern',
        score: weightedMean(bucket.schoolScores.Northwestern, 'adjustedNorthwestern', 'weight'),
        criteriaScored: bucket.criteriaScored,
        totalCriteria: bucket.totalCriteria,
      },
    ]
  })

  const drivers = pairedRows
    .slice()
    .sort((left, right) => Math.abs(right.contributionGap) - Math.abs(left.contributionGap))
    .slice(0, 10)
    .map((row) => ({
      criterionId: row.criterionId,
      label: row.label,
      leadingSchool: row.scoreGap >= 0 ? 'UCSB' : 'Northwestern',
      scoreGap: Math.abs(row.scoreGap),
      contributionGap: Math.abs(row.contributionGap),
    }))

  const missing = frame.filter((row) => !row.paired)

  return {
    overall,
    coverage: frame.length === 0 ? 0 : pairedRows.length / frame.length,
    categorySummary,
    drivers,
    missing,
    detailRows: frame,
  }
}

export function createMarkdownReport(comparison) {
  const overall = comparison.overall
  const winner = overall.reduce((best, current) => {
    if (!best || (current.overallScore ?? -Infinity) > (best.overallScore ?? -Infinity)) {
      return current
    }
    return best
  }, null)

  const lines = [
    '# Camille Decision Report',
    '',
    `Generated on ${new Date().toISOString()}.`,
    '',
    'This report combines research-baseline scores for objective criteria with prefilled or edited survey scores for subjective criteria.',
    '',
    '## Overall Recommendation',
    '',
    winner ? `Current recommendation: **${winner.school}** has the higher weighted overall score.` : 'No recommendation available.',
    '',
    ...overall.map((row) => `- ${row.school}: ${row.overallScore?.toFixed(2) ?? 'NA'}`),
    '',
    `Coverage: ${(comparison.coverage * 100).toFixed(1)}%`,
    '',
    '## Category Breakdown',
    '',
    '| Category | School | Score | Criteria Scored | Total Criteria |',
    '| --- | --- | ---: | ---: | ---: |',
    ...comparison.categorySummary.map(
      (row) =>
        `| ${row.category.replaceAll('_', ' ')} | ${row.school} | ${row.score?.toFixed(2) ?? '-'} | ${row.criteriaScored} | ${row.totalCriteria} |`,
    ),
    '',
    '## Top Drivers',
    '',
    ...(comparison.drivers.length > 0
      ? comparison.drivers.map(
          (driver) =>
            `- ${driver.leadingSchool} leads on **${driver.label}** by ${driver.scoreGap.toFixed(2)} adjusted score points.`,
        )
      : ['- No strong driver list is available yet.']),
    '',
    '## Criterion-Level Detail',
    '',
    '| Category | Criterion | Weight | UCSB Score | UCSB Source | Northwestern Score | Northwestern Source |',
    '| --- | --- | ---: | ---: | --- | ---: | --- |',
    ...comparison.detailRows.map(
      (row) =>
        `| ${row.category.replaceAll('_', ' ')} | ${row.label} | ${row.weight.toFixed(1)} | ${row.ucsbScore.toFixed(1)} | ${row.ucsbSource} | ${row.northwesternScore.toFixed(1)} | ${row.northwesternSource} |`,
    ),
    '',
  ]

  return lines.join('\n')
}
